clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -6300,7 +6300,7 @@ NamedDecl *Sema::FindInstantiatedDecl(SourceLocation Loc, NamedDecl *D,
                   getTrivialTemplateArgumentLoc(UnpackedArg, QualType(), Loc));
           }
           QualType T = CheckTemplateIdType(TemplateName(TD), Loc, Args);
-          if (T.isNull())
+          if (T.isNull() || T->containsErrors())
```

Comment:
Is there a reason this check isn’t done in `CheckTemplateIdType()` instead? Because there are a few other places where this exact pattern appears; is there a difference between this and other uses of `CheckTemplateIdType()` that makes it so it only makes sense to check for errors here (and if so, then that would probably deserve a comment explaining that)?

Reply:
There is no reason besides that it is where I saw the crash. I have moved into `CheckTemplateIdType`.

Reply:
I don't understand the logic of this change: If there is no reason to make `CheckTemplateIdType` fail for more cases, why make this change then?

Usually in error recovery, we want to carry the error information as much as we can, where practical.
In this case, there is the same implementation complexity either way.
This change went in the opposite direction: We are bailing out earlier for more cases, and we have no reason for it.

Reply:
I have checked the other places where `CheckTemplateIdType` is used, and indeed, I can not (easily, at least) find any reason why we would have a crash in those places. For instance, in [`SemaTemplate.cpp`](https://github.com/llvm/llvm-project/blob/9ae24c9ac94017e15eb827e25c5693418e6cdb4b/clang/lib/Sema/SemaTemplate.cpp#L5072) there is a similar pattern (as @Sirraide was concerned),  but there a null `RecordType` is appropriately handled.

While looking at it, I realized that the null pointer dereference we had fixed downstream was fixed upstream [four months ago](https://github.com/llvm/llvm-project/commit/7415524b45392651969374c067041daa82dc89e7#diff-761c59b5215d832845bd988566c7d0abaa1cc6294cb8ea90f32e33d15bfccf28R6290) in #77890. The introduced assertions trigger, though, throwing me off (since the stack trace is almost identical).

I have gone back to the original commit and added a comment. I will also rework the PR title and description.

Reply:
Hmm, yeah, if every other place where this is used can already deal with a type that contains errors, then that makes sense.